eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Utxo.hs

```diff
@@ -1,58 +1,403 @@
+      !(NonEmpty (TxIn (EraCrypto era)))
+  deriving (Generic)
+
+type instance EraRuleFailure "UTXO" (ConwayEra c) = ConwayUtxoPredFailure (ConwayEra c)
 
 type instance EraRuleEvent "UTXO" (ConwayEra c) = AlonzoUtxoEvent (ConwayEra c)
 
-instance InjectRuleFailure "UTXO" BabbageUtxoPredFailure (ConwayEra c)
+instance InjectRuleFailure "UTXO" ConwayUtxoPredFailure (ConwayEra c)
+
+instance InjectRuleFailure "UTXO" BabbageUtxoPredFailure (ConwayEra c) where
+  injectFailure = babbageToConwayUtxoPredFailure
 
 instance InjectRuleFailure "UTXO" AlonzoUtxoPredFailure (ConwayEra c) where
-  injectFailure = AlonzoInBabbageUtxoPredFailure
+  injectFailure = alonzoToConwayUtxoPredFailure
 
 instance InjectRuleFailure "UTXO" ShelleyUtxoPredFailure (ConwayEra c) where
   injectFailure =
-    AlonzoInBabbageUtxoPredFailure
-      . allegraToConwayUtxoPredFailure
+    allegraToConwayUtxoPredFailure
       . shelleyToAllegraUtxoPredFailure
 
 instance InjectRuleFailure "UTXO" Allegra.AllegraUtxoPredFailure (ConwayEra c) where
-  injectFailure = AlonzoInBabbageUtxoPredFailure . allegraToConwayUtxoPredFailure
+  injectFailure = allegraToConwayUtxoPredFailure
 
 instance InjectRuleFailure "UTXO" ConwayUtxosPredFailure (ConwayEra c) where
-  injectFailure = AlonzoInBabbageUtxoPredFailure . UtxosFailure
+  injectFailure = UtxosFailure
 
 instance InjectRuleFailure "UTXO" AlonzoUtxosPredFailure (ConwayEra c) where
-  injectFailure = AlonzoInBabbageUtxoPredFailure . UtxosFailure . injectFailure
+  injectFailure =
+    alonzoToConwayUtxoPredFailure
+      . Alonzo.UtxosFailure
+      . injectFailure
+
+deriving instance
+  ( Era era
+  , Show (Value era)
+  , Show (PredicateFailure (EraRule "UTXOS" era))
+  , Show (TxOut era)
+  , Show (Script era)
+  , Show (TxIn (EraCrypto era))
+  ) =>
+  Show (ConwayUtxoPredFailure era)
+
+deriving instance
+  ( Era era
+  , Eq (Value era)
+  , Eq (PredicateFailure (EraRule "UTXOS" era))
+  , Eq (TxOut era)
+  , Eq (Script era)
+  , Eq (TxIn (EraCrypto era))
+  ) =>
+  Eq (ConwayUtxoPredFailure era)
+
+instance
+  ( Era era
+  , NFData (Value era)
+  , NFData (TxOut era)
+  , NFData (PredicateFailure (EraRule "UTXOS" era))
+  ) =>
+  NFData (ConwayUtxoPredFailure era)
+
+--------------------------------------------------------------------------------
+-- ConwayUTXO STS
+--------------------------------------------------------------------------------
+
+instance
+  forall era.
+  ( EraTx era
+  , EraUTxO era
+  , ConwayEraTxBody era
+  , AlonzoEraTxWits era
+  , Tx era ~ AlonzoTx era
+  , EraRule "UTXO" era ~ ConwayUTXO era
+  , InjectRuleFailure "UTXO" ShelleyUtxoPredFailure era
+  , InjectRuleFailure "UTXO" AllegraUtxoPredFailure era
+  , InjectRuleFailure "UTXO" AlonzoUtxoPredFailure era
+  , InjectRuleFailure "UTXO" BabbageUtxoPredFailure era
+  , InjectRuleFailure "UTXO" ConwayUtxoPredFailure era
+  , -- instructions for calling UTXOS from BabbageUTXO
+    Embed (EraRule "UTXOS" era) (ConwayUTXO era)
```

Comment:
That was a pretty useless comment for Babbage, but here it makes even less sense.
```suggestion
  , Embed (EraRule "UTXOS" era) (ConwayUTXO era)
```